vmsyscall/src/bootinfo.rs

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! The `bootinfo` module includes the hard-coded memory layout and
+//! the `BootInfo` structure which is handed to the kernel.
+//!
+//! Initially copied from:
+//! https://github.com/rust-osdev/bootloader/blob/90f5b8910d146d6d489b70a6341d778253663cfa/src/bootinfo/mod.rs
+
+use crate::memory_map::Map;
+use core::fmt;
+
+/// This structure represents the information that the bootloader passes to the kernel.
+///
+/// The information is passed as an argument to the entry point:
+///
+/// ```ignore
+/// pub extern "C" fn _start(boot_info: &'static BootInfo) -> ! {
+///    // […]
+/// }
+/// ```
+///
+/// Note that no type checking occurs for the entry point function, so be careful to
+/// use the correct argument types. To ensure that the entry point function has the correct
+/// signature, use the [`entry_point`] macro.
+#[derive(Clone)]
+#[repr(C)]
+pub struct BootInfo {
+    /// A map of the physical memory regions of the underlying machine.
+    ///
+    /// The bootloader queries this information from the BIOS/UEFI firmware and translates this
+    /// information to Rust types. It also marks any memory regions that the bootloader uses in
+    /// the memory map before passing it to the kernel. Regions marked as usable can be freely
+    /// used by the kernel.
+    pub memory_map: Map,
+
+    /// ELF entry point of the ring 3 executable
+    pub entry_point: *const u8,
+
+    /// ELF program headers of the ring 3 executable
+    pub load_addr: *const u8,
+
+    /// ELF number of program headers of the ring 3 executable
+    pub elf_phnum: usize,
```

Comment:
We only support `static-pie` binaries. If the kernel passes the (randomized) code address to the host, the remaining items can be parsed from the ELF header (trivially).

Reply:
Do you mean, that `load_addr` which basically is the address of [ElfHeader64](https://docs.rs/goblin/0.2.1/goblin/elf/header/struct.Header.html) already contains `entry_point` and `elf_phnum`? If so... yes.
Yesterday I implemented ASLR for the ring3 `code` with this information.
The loader can load the code anywhere in this case. The shim can then construct the page tables with ASLR.